Kicking off a WebGL implementation

dmarcos opened the pull request:
@jdm @ecoal95 I'm working on making VR happen in the Browser and I want to bring to Servo the [webVR APIs](https://github.com/MozVR/webvr-spec/blob/master/webvr.idl) we already have in Gecko. Before anything happens we need a working implementation of WebGL (and also the [fullscreen API](https://fullscreen.spec.whatwg.org/)). My implementation is very basic and probably naive (I just recently started to contribute to Servo). My patch is just a starting point: 
- It only implements `clearColor` and `clear` methods of the [WebGL spec](https://www.khronos.org/registry/webgl/specs/latest/).
- It uses the readback strategy that `canvasRenderingContext2D` is using (The webgl task paints stuff independently on it's own buffer and the compositor task request the pixels back to the webgl task when it needs them) I'm sure there are much better ways to handle this. Latency and FPS are critical in VR so we have to figure out the fastest way to push pixels to the screen. I've read something about layerizing the canvas but I'm still not sure what that even means :)
- There's an included test you can try `./mach run tests/ref/webgl-context/clearcolor.html` 

@ecoal95 I know you'll be working on this for the next three months. With a foundation in place we will be able to make quick progress in parallel. This is exciting!

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/servo/servo/5652)

<!-- Reviewable:end -->

dmarcos committed:
Kicks off a WebGL implementation

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!

ecoal95 commented:
@dmarcos Wow, that's awesome!

I'm currently working in getting a hardware-accelerated 3d rendering context here: https://github.com/ecoal95/rust-offscreen-rendering-context

I started implementing WebGLRenderingContext in servo locally yesterday... But I don't have your progress by far so... Using this codebase will be great! :P

As far as I'm concerned, layerizing the canvas is moving it to its own rendering layer, so it can be associated with a `RenderContext`, which be associated to the same shared surface attached to the native 3D context associated to the `WebGLRenderingContext`, so rendering the canvas will use directly data from GPU, and thus avoiding readback. I don't know if I'm right here, still learning a lot ;)

Probably @pcwalton will know if I'm right.

I have some notes about your code here:
  * I'd use experimental-webgl instead of webgl right now, but that's trivial to change.
  * What about separating the CanvasMsg between 3D messages and 2D ones? I know readback won't be the final implementation but...
  * When you use read_pixels, we should be able to pass gl::BGRA to it. Unless there's any implication I don't know of, I could make a PR to gleam to allow it, am I right? The same counterpart here: read_pixels will be temporary

Nice to meet you! It's exciting to work with you :)

dmarcos commented:
@ecoal95 
- We can use both "webgl" and "webgl-experimental".
- I wanted to minimize the changes in other parts of the code (display_list_builder.rs) and a combined CanvasMsg for 2d and webgl was an easier start. 
- gl:BGRA is not supported on gleam. Feel free to do a PR on gleam and we can get rid of the byte swapping.

pcwalton commented:
I'll take this review.

pcwalton commented on components/canvas/canvas_msg.rs:
```diff
@@ -0,0 +1,46 @@
+
+#[derive(Clone)]
+pub enum CanvasMsg {
+    // 2dContext
```
Please move this to be a doc comment on `CanvasMsg`:

    /// 2D context.
    #[derive(Clone)]
    pub enum CanvasMsg {
        ...
    }

pcwalton commented on components/canvas/canvas_msg.rs:
```diff
@@ -0,0 +1,46 @@
+    GetImageData(Rect<f64>, Size2D<f64>, Sender<Vec<u8>>),
+    PutImageData(Vec<u8>, Rect<f64>, Option<Rect<f64>>),
+    Close,
+    // WebGL
```
Oh, I see what you're doing. Can you split this up into three enums: `Canvas2dMsg`, `CanvasWebGlMsg`, and `CommonCanvasMsg`? Then you can just have `CanvasMsg { Canvas2d(Canvas2dMsg), WebGl(CanvasWebGlMsg), Common(CommonCanvasMsg) }`.

pcwalton commented on components/canvas/canvas_paint_task.rs:
```diff
@@ -255,6 +226,7 @@ impl<'a> CanvasPaintTask<'a> {
                     CanvasMsg::PutImageData(imagedata, image_data_rect, dirty_rect)
                         => painter.put_image_data(imagedata, image_data_rect, dirty_rect),
                     CanvasMsg::Close => break,
+                    _ => panic!("Wrong message sent to Canvas2D task"),
```
This is why multiple enums are better: this catchall will suppress nonexhaustive-match compiler warnings.

pcwalton commented on components/canvas/webgl_paint_task.rs:
```diff
@@ -0,0 +1,155 @@
+            loop {
+                match port.recv().unwrap() {
+                    CanvasMsg::Clear(mask) => {
+                      painter.clear(mask);
```
Please use 4-space indents.

pcwalton commented on components/canvas/webgl_paint_task.rs:
```diff
@@ -0,0 +1,155 @@
+        let draw_target = DrawTarget::new_with_fbo(BackendType::Skia,
+                                                   graphics_context,
+                                                   size,
+                                                   SurfaceFormat::B8G8R8A8);
```
There's no need to create a draw target here, since WebGL has no vector operations. Just store a blob of pixels for now and create the source surface in the paint task. (Eventually we will want to layerize.)

pcwalton replied:
Also definitely add a `FIXME` and file an issue to layerize here.

pcwalton commented on components/script/dom/htmlcanvaselement.rs:
```diff
@@ -74,8 +91,22 @@ pub trait LayoutHTMLCanvasElementHelpers {
-        let context = (*self.unsafe_get()).context.get_inner_as_layout();
-        context.map(|cx| cx.get_renderer())
+        let ref canvas = *self.unsafe_get();
+        match canvas.context_2d.get() {
```
if canvas.context_2d.get().is_some() {
        ...
    } else if canvas.context_webgl.get().is_some() {
        ...
    } else {
        None
    }

pcwalton commented on components/script/dom/htmlcanvaselement.rs:
```diff
@@ -101,7 +133,23 @@ impl<'a> HTMLCanvasElementHelpers for JSRef<'a, HTMLCanvasElement> {
+        let context = self.GetContext(String::from_str("2d"));
+        match context.unwrap() {
+          CanvasRenderingContext2DOrWebGLRenderingContext::eCanvasRenderingContext2D(context) => {
+              return Temporary::new(context.root().r().unrooted());
```
No need for `return` or the semicolon here

pcwalton commented on components/script/dom/htmlcanvaselement.rs:
```diff
@@ -101,7 +133,23 @@ impl<'a> HTMLCanvasElementHelpers for JSRef<'a, HTMLCanvasElement> {
-        self.GetContext(String::from_str("2d")).unwrap()
+        let context = self.GetContext(String::from_str("2d"));
+        match context.unwrap() {
+          CanvasRenderingContext2DOrWebGLRenderingContext::eCanvasRenderingContext2D(context) => {
```
Use 4-space indents.

pcwalton commented on components/script/dom/htmlcanvaselement.rs:
```diff
@@ -128,17 +176,29 @@ impl<'a> HTMLCanvasElementMethods for JSRef<'a, HTMLCanvasElement> {
+               });
+               return Some(CanvasRenderingContext2DOrWebGLRenderingContext::eCanvasRenderingContext2D(Unrooted::from_temporary(context_2d)));
+           }
+           "webgl" => {
```
Support `experimental-webgl` as well.

pcwalton commented on components/script/dom/htmlcanvaselement.rs:
```diff
@@ -128,17 +176,29 @@ impl<'a> HTMLCanvasElementMethods for JSRef<'a, HTMLCanvasElement> {
+                   let (w, h) = (self.width.get() as i32, self.height.get() as i32);
+                   CanvasRenderingContext2D::new(GlobalRef::Window(window.r()), self, Size2D(w, h))
+               });
+               return Some(CanvasRenderingContext2DOrWebGLRenderingContext::eCanvasRenderingContext2D(Unrooted::from_temporary(context_2d)));
```
No need for `return`. (This applies to the other branches too.)

pcwalton commented on components/script/dom/htmlcanvaselement.rs:
```diff
@@ -60,6 +65,18 @@ impl HTMLCanvasElement {
+
+    fn recreate_contexts(&self) {
+        let (w, h) = (self.width.get() as i32, self.height.get() as i32);
+        match self.context_2d.get() {
```
Try `if let`:

    if let Some(context) = self.context_2d.get() {
        context.root().r().recreate(Size2D(w, h))
    }

pcwalton commented on components/script/dom/htmlcanvaselement.rs:
```diff
@@ -60,6 +65,18 @@ impl HTMLCanvasElement {
     }
+
+    fn recreate_contexts(&self) {
+        let (w, h) = (self.width.get() as i32, self.height.get() as i32);
```
I've seen this in several places. Why not factor out this into an `fn size()` that returns a `Size2D<i32>`?

pcwalton commented on components/script/dom/htmlcanvaselement.rs:
```diff
@@ -60,6 +65,18 @@ impl HTMLCanvasElement {
+            Some(context) => context.root().r().recreate(Size2D(w, h)),
+            None => ()
+        }
+        match self.context_webgl.get() {
```
`if let` here too.

pcwalton commented on components/canvas/webgl_paint_task.rs:
```diff
@@ -0,0 +1,155 @@
+        let source_surface = self.drawtarget.create_source_surface_from_data(&pixels,
+            self.size, self.size.width * 4, SurfaceFormat::B8G8R8A8);
+
+        let canvas_rect = Rect(Point2D(0i32, 0i32), self.size);
```
nit: I don't think you will need `0i32`, just write `0`.

pcwalton commented on components/canvas/webgl_paint_task.rs:
```diff
@@ -0,0 +1,155 @@
+
+        let canvas_rect = Rect(Point2D(0i32, 0i32), self.size);
+        let draw_surface_options = DrawSurfaceOptions::new(Filter::Linear, true);
+        let draw_options = DrawOptions::new(1.0f64 as AzFloat, 0);
```
nit: `1.0` is sufficient (but I don't think you need this code at all, see above).

pcwalton commented on tests/ref/webgl-context/clearcolor_ref.html:
```diff
@@ -0,0 +1,20 @@
+<html>
+<head>
+  <meta charset="utf-8">
+  <title>WegGL ClearColor Test</title>
```
typo: "WebGL"

pcwalton commented on components/script/dom/htmlcanvaselement.rs:
```diff
@@ -192,12 +248,9 @@ impl<'a> VirtualMethods for JSRef<'a, HTMLCanvasElement> {
+            self.recreate_contexts();
         }
     }
+
```
nit: Remove this blank line

pcwalton commented on components/script/dom/htmlcanvaselement.rs:
```diff
@@ -128,17 +176,29 @@ impl<'a> HTMLCanvasElementMethods for JSRef<'a, HTMLCanvasElement> {
+               return Some(CanvasRenderingContext2DOrWebGLRenderingContext::eWebGLRenderingContext(Unrooted::from_temporary(context_webgl)));
+           }
+           _ => return None
+
```
nit: Remove this blank line

pcwalton commented:
Awesome, this is a great start! Apologies for the deluge of comments ):

dmarcos commented:
@pcwalton Thanks for finding the time to review this. I addressed your comments and open an issue (https://github.com/servo/servo/issues/5679) to layerize the canvas. Is there anything else you want me to do?

bors-servo commented:
:umbrella: The latest upstream changes (presumably #5680) made this pull request unmergeable. Please resolve the merge conflicts.

pcwalton commented on components/canvas/webgl_paint_task.rs:
```diff
@@ -0,0 +1,119 @@
+    fn create(size: Size2D<i32>,
+              native_graphics_context: &NativePaintingGraphicsContext) {
+        let graphics_context = native_graphics_context as *const _ as SkiaGrGLNativeContextRef;
+        // It creates an OpenGL context
```
This is fine for now, but calling our Skia extension to do this is pretty ugly. Could you add a `FIXME` saying "provide a clean, cross-platform way to do this, e.g. in glutin"?

pcwalton commented on components/canvas/webgl_paint_task.rs:
```diff
@@ -0,0 +1,119 @@
+    }
+
+    fn send_pixel_contents(&mut self, chan: Sender<Vec<u8>>) {
+        // FIXME: Instead of a readback strategy we have
```
Add the issue number here.

pcwalton commented on components/script/dom/htmlcanvaselement.rs:
```diff
@@ -90,18 +117,35 @@ impl LayoutHTMLCanvasElementHelpers for LayoutJS<HTMLCanvasElement> {
-    }
+    // fn get_size(&self) -> Size2D<i32> {
+    //     Size2D(self.Width() as i32, self.Height() as i32)
+    // }
```
Delete this if no longer used.

pcwalton commented on components/script/dom/htmlcanvaselement.rs:
```diff
@@ -90,18 +117,35 @@ impl LayoutHTMLCanvasElementHelpers for LayoutJS<HTMLCanvasElement> {
 
 pub trait HTMLCanvasElementHelpers {
-    fn get_size(&self) -> Size2D<i32>;
+    //fn get_size(&self) -> Size2D<i32>;
```
Delete this if no longer used.

pcwalton commented:
Almost there! Just a few nits to address and looks great otherwise :)

dmarcos commented:
@pcwalton Thanks again for helping with this! I rebased and addressed your comments. Do you want me to squash before merging?

pcwalton commented:
@dmarcos Yup, squash away!

dmarcos commented:
@pcwalton Squashed and rebased! Ready to ship :)

jdm commented:
@bors-servo: r=pcwalton

bors-servo commented:
:pushpin: Commit c448988 has been approved by `pcwalton`

<!-- @bors-servo r=pcwalton c448988b2f815b45dafaa33016a8f8cf76842ede -->

bors-servo commented:
:hourglass: Testing commit c448988 with merge e7549e8...

bors-servo commented:
:broken_heart: Test failed - [gonk](http://build.servo.org/builders/gonk/builds/608)

pcwalton commented:
Perhaps the gonk Cargo.lock wasn't updated?

dmarcos commented:
@pcwalton I updated Cargo.lock to the latest hash of gleam and pushed again

pcwalton commented:
@bors-servo: r+

bors-servo commented:
:pushpin: Commit d3669b2 has been approved by `pcwalton`

<!-- @bors-servo r=pcwalton d3669b2ef9967a635367009604a00f8492862e8b -->

bors-servo commented:
:hourglass: Testing commit d3669b2 with merge da528cd...

bors-servo commented:
:broken_heart: Test failed - [mac2](http://build.servo.org/builders/mac2/builds/1652)

jdm commented:
https://github.com/servo/gleam/commit/228ad25671ddaca02c53ca0177c50ccb17f8b868 updated gleam past the version of rustc that we're using, unfortunately.

dmarcos commented:
@jdm Just pushed a change to make Cargo.lock point to https://github.com/servo/gleam/commit/228ad25671ddaca02c53ca0177c50ccb17f8b868

jdm commented:
I don't think that will work, for the reason that I gave in my last comment :)

dmarcos commented:
@jdm What version should I use?

dmarcos commented:
@jdm Updated gleam to https://github.com/servo/gleam/commit/70c5f2ea3ef3602277b7c380c24b504e42338056. Will this do the trick?

jdm commented:
We probably need to make a branch with that commit applied on top of the current revision we're using.

ecoal95 commented:
@dmarcos The other option is going back to the old gleam version and use an unsafe `set_len` after `read_pixels` to bypass this until gleam gets updated, its the only thing that won't work.

jdm commented:
That sounds easier to deal with when upgrading Servo's rustc later; I like it.

ecoal95 commented:
I use it here: https://github.com/ecoal95/servo/blob/webgl/components/canvas/webgl_paint_task.rs#L85

Using a FIXME or a TODO should be a good idea.

dmarcos commented:
@ecoal95 I don't think you can just change the size of the array. You are going beyond the boundaries of the array originally allocated(https://github.com/servo/gleam/blob/master/src/gl.rs#L77) and it won't contain valid pixel values.

dmarcos commented:
@jdm Don't you think @jdm Updated gleam to https://github.com/servo/gleam/commit/70c5f2ea3ef3602277b7c380c24b504e42338056 will work? It's a commit pre https://github.com/servo/gleam/commit/228ad25671ddaca02c53ca0177c50ccb17f8b868

ecoal95 commented:
Yes, I'm aware of it, but when we change the length OpenGL has already written on it, so if it fails it fails on gleam's code. It's just a workaround until the fixed gleam version comes out.

dmarcos commented:
@jdm Can we let buildbot speak? :)
If https://github.com/servo/gleam/commit/228ad25671ddaca02c53ca0177c50ccb17f8b868 is the commit tha breaks the build https://github.com/servo/gleam/commit/70c5f2ea3ef3602277b7c380c24b504e42338056 should work and it has the fix we need

jdm commented:
Worth a shot!

jdm commented:
@bors-servo: r+

bors-servo commented:
:pushpin: Commit 464d067 has been approved by `jdm`

<!-- @bors-servo r=jdm 464d067acbe797c3eed63e675701f194fbf685d7 -->

bors-servo commented:
:hourglass: Testing commit 464d067 with merge e4fd2dc...

bors-servo commented:
:broken_heart: Test failed - [gonk](http://build.servo.org/builders/gonk/builds/613)

jdm commented:
You'll need to run `./mach update-cargo -p gleam --precise [hash]` to ensure all the .lock files are updated.

dmarcos commented:
@jdm Here we go! I run your magic command and pushed again. Crossing my fingers

jdm commented:
@bors-servo: r+

bors-servo commented:
:pushpin: Commit 4bcc3cd has been approved by `jdm`

<!-- @bors-servo r=jdm 4bcc3cd6942364cbb52b150cb739509e0a94f654 -->

bors-servo commented:
:hourglass: Testing commit 4bcc3cd with merge fb649c4...

bors-servo commented:
:broken_heart: Test failed - [linux1](http://build.servo.org/builders/linux1/builds/1442)

dmarcos commented:
My ref tests fails on linux1. Investigating now

dmarcos commented:
@jdm Not sure what to do here. This works fine on OSX 

![Ok Test] (http://f.cl.ly/items/1O0X2e1u2g021K2o1u1Q/Image%202015-04-15%20at%2011.48.11%20AM.png)



The buildbot output is not giving me much info and I don't have a Linux machine to investigate. Should I just disable the ref test?

jdm commented:
Let's find someone who can help. @glennw, can you take a look?

dmarcos commented:
@glennw You can run ```./mach run tests/ref/webgl-context/clearcolor.html``` and ```./mach test tests/ref/webgl-context```

dmarcos commented:
@glennw It's this way of creating a gl context a problem on Linux?

https://github.com/servo/servo/pull/5652/files#diff-c7c3a12b3d29da08ebeceaa8341702d2R69

ecoal95 commented:
Just tested on my Ubuntu 14.10 with Intel drivers and passed the test.

![screenshot from 2015-04-16 00 18 57](https://cloud.githubusercontent.com/assets/1323194/7170437/6cf49192-e3ce-11e4-82ca-45130dcf7a0f.png)

dmarcos commented:
Thanks @ecoal95. I just pushed a commit that uses glutin instead of skia to create the gl context. I don't know if it will make any difference on buildbot but the code is cleaner now. @jdm Can you take a look to the last commit? If you think it's okay we can give another shot on buildbot.

ecoal95 commented:
Anyways, I wouldn't complicate it too much, since we'll need custom context creation to use hardware acceleration when possible, to allow choosing capabilities via WebGLContextAttributes, and probably to layerize too.

dmarcos commented:
@jdm I updated the glutin used by canvas to the servo branch. When you give me the ok I'll squash and we can send it to buildbot. @ecoal95 tried on Linux and the test passed for him. Is there anything particular in the buildbot environment that might be making the test fail?

jdm commented:
The full buildbot configuration should live in https://github.com/servo/saltfs/blob/master/buildbot/master/master.cfg but I don't know of anything off the top of my head. Go ahead and squash these and we can try again.

dmarcos commented:
@jdm Squashed and pushed

jdm commented:
@bors-servo: r+

bors-servo commented:
:pushpin: Commit 85b3937 has been approved by `jdm`

<!-- @bors-servo r=jdm 85b39370a47c0541e4676e26ec0dcedb55dc33a1 -->

bors-servo commented:
:hourglass: Testing commit 85b3937 with merge 30a1c75...

bors-servo commented:
:broken_heart: Test failed - [gonk](http://build.servo.org/builders/gonk/builds/624)

dmarcos commented:
@jdm Mmmh, What is this? It doesn't seem related to my changes

larsbergstrom commented:
@dmarcos I'm investigating that failure locally.

dmarcos commented:
@larsbergstrom Thanks! Do you want me to rebase and push again?

larsbergstrom commented:
@dmarcos I think we're OK for right now - GH thinks it can merge fine. If it bitrots, feel free to go ahead and do so, though. I should be able to repro the failure either way, hopefully!

larsbergstrom commented:
When I perform a gonk build locally, it updates ports/cef/Cargo.lock with an updated android_glue, which is where I suspect the change may be coming from. Continuing to investigate...

larsbergstrom commented:
Hrm, where the old system would use `-mandroid`, we now get several calls to `-landroid`. The plot thickens...

dmarcos commented:
@larsbergstrom Do you think is it something I've done or it comes from somewhere else.

dmarcos commented:
@jdm Just rebased and pushed and the warning is gone. Can we show it to buildbot to see whether it likes it?

larsbergstrom commented:
@dmarcos We can give it another retry, but I suspect it will still fail on gonk+android. I can't quite track down where in the world the new "-landroid" is coming from. The only new place I see it is in git/checkouts/rust-egl-995f9207ca3ade64/master/test/egl-android-glue/jni/Android.mk

dmarcos commented:
@larsbergstrom I'm using glutin in my patch and as requested by @jdm I made sure we were using the servo branch:

https://github.com/servo/servo/pull/5652/files#diff-3e8ba4385f421ecb49eebac210f90118R23

Could this be triggering any sort of change in the sub dependencies?

dmarcos commented:
@jdm Should we give this another shot?

jdm commented:
@dmarcos I'm not sure why the outcome would be any different than before; what change have you made?

dmarcos commented:
@jdm I had observed there was a warning in my machine affer doing ```./mach clean``` I rebased and the warning is gone. I also removed two obsolete dependencies from the canvas/Cargo.toml (compositor and layers) I don't understand how the dependency tree is managed and created. I would like to see what buildbot says.

jdm commented:
Want to get rid of skia as well?

dmarcos commented:
@jdm I removed skia from canvas/Cargo.toml but remained in the Cargo.lock files. What am I supposed to do to make sure that they are updated? Do I have to remove the skia references manually?

larsbergstrom commented:
I am assuming this is related to the fact that when I build the Cargo.lock file adds:
```
+name = "android_glue"
+version = "0.0.1"
+source = "git+https://github.com/tomaka/android-rs-glue#a0b251d5533475f36d85583d92a0feed58f09817"
+
```

dmarcos commented:
@jdm Okay. I removed compositor / layers and skia references from Cargo.locks. ```./mach clean``` + ```./mach build``` finishes clean with no warnings: https://pastebin.mozilla.org/8830362

There's nothing left in the patch that it's superfluous. I would give another shot to buildbot. If it doesn't work I will let the experts carry on :)

@larsbergstrom I don't see this ```name = "android_glue"...``` on my Cargo.locks. I'm on OSX.

larsbergstrom commented:
Aha, here it is! https://github.com/tomaka/android-rs-glue/blame/master/glue/src/ffi.rs#L77

We're picking up a new version of that library via glutin.

dmarcos commented:
@larsbergstrom The version of glutin I'm using is the servo branch

jdm commented:
@dmarcos: It looks like the glutin dependency hasn't been added to the Cargo.lock files under ports/, so the build machines end up picking up the most recent one and causing the build failure. Try running the appropriate `./mach update-cargo` command again?

dmarcos commented:
I've run ./mac update-cargo and I get when building ```native library `openssl` is being linked to by more than one package, and can only be linked to by one package```

jdm commented:
Right, I meant `./mach update-cargo -p glutin --precise 0feab4842c5b229bfe88739d2c526d03c198fd33`. If that doesn't cause the Cargo.lock files to change, you may need to run this sequence:
```
./mach build-cef
[ctr-c]
GONKDIR=/tmp ./mach build-gonk
[ctr-c]
```
before re-running the earlier update-cargo.

larsbergstrom commented:
@bors-servo: retry

dmarcos commented:
@larsbergstrom @jdm It looks more reasonable I see glutin mentioned in the ```ports/.../Cargo.lock" files. There's another bunch of changes that I don't understand:

https://github.com/servo/servo/commit/2af2ab9f2129b212dba31c1a8dbce7aa7d50c5ec#diff-60204fa1bd8b7a5c8e7e1e0705ac7542L10

Let me know if you want me to change anything

jdm commented:
@bors-servo: r+

bors-servo commented:
:pushpin: Commit f72753e has been approved by `jdm`

<!-- @bors-servo r=jdm f72753e92a90225143fc67bfc4e7fd59bf6e975b -->

bors-servo commented:
:hourglass: Testing commit f72753e with merge 883a244...

larsbergstrom commented:
@dmarcos after the failed `update-cargo` without `--precise`, did you `git checkout -- ` the lockfiles? That many version updates looks suspiciously like cargo running amok trying to update everything in the world and failing.

dmarcos commented:
@larsbergstrom Yes, I had reverted the lockfiles to upstream/master ```git checkout upstream/master port/../Cargo.lock```. If the current diff doesn't look reasonable to you I can do it again:

https://github.com/servo/servo/pull/5652/files

jdm commented:
It's entirely possible that nobody has built gonk locally and committed the changes to its Cargo.lock since the last upgrade, which could result in this.

bors-servo commented:
:broken_heart: Test failed - [gonk](http://build.servo.org/builders/gonk/builds/640)

dmarcos commented:
@larsbergstrom Still warnings

dmarcos commented:
@jdm Any idea on how I can make progress with this? I cannot wait to merge this patch :)

jdm commented:
I propose:
* reset all Cargo.lock changes in all directories
* `./mach update-cargo -p glutin --precise 0feab4842c5b229bfe88739d2c526d03c198fd33`
* `/mach update-cargo -p gleam --precise 70c5f2e`
* `./mach build`, interrupt it at the first "compiling"
* `./mach build-cef`, interrupt it at the first "compiling
* `GONKDIR=/tmp ./mach build-gonk`, interrupt it at the first "compiling"
* commit and push the result

dmarcos commented:
@jdm Thanks! Done! You can see the complete sequence of commands I followed if you need:

https://pastebin.mozilla.org/8830417

jdm commented:
Oh, I think I see a problem - the new android_glue that gonk is using is rev a0b251d5533475f36d85583d92a0feed58f09817, but the one that servo currently uses elsewhere is 5a68056599fb498b0cf3715fd359894825e3b856. Try using `./mach update-cargo -p android_glue --precise 5a68056599fb498b0cf3715fd359894825e3b856`?

dmarcos commented:
@jdm This is what I get:
```bash

./mach update-cargo -p android_glue --precise 5a68056599fb498b0cf3715fd359894825e3b856
components/servo
There are multiple `android_glue` packages in your project, and the specification `android_glue` is ambiguous.
Please re-run this command with `-p <spec>` where `<spec>` is one of the following:
  file:///Users/dmarcos/Development/VR/servo/components/servo#android_glue:0.0.1
  https://github.com/tomaka/android-rs-glue#android_glue:0.0.1

```

larsbergstrom commented:
@dmarcos I believe that other one is coming from the submodule (support/android-rs-glue), which we require because of some cargo limitation that I've long forgotten (though @glennw might remember why we have it?).

I'm trying:
```
cd support/android-rs-glue
git fetch upstream
git checkout 5a68056599fb498b0cf3715fd359894825e3b856
cd ..
git add android-rs-glue
```
and doing a rebuild of the gonk configuration locally to see if that fixes it.

I apologize for all of the cargo horrors you've gone through trying to land this!

dmarcos commented:
@larsbergstrom No problem. Let's land this beast! :)

larsbergstrom commented:
UGH. It looks like even if we fix android-rs-glue, the glutin update brings in a dependency on libandroid.so:
```
larsberg@ubuntu:~/servo$ find . -name '*.rs' | xargs grep -e "\"android\"" | grep link
./.cargo/git/checkouts/glutin-72cd277ab8976bc1/servo/src/android/ffi.rs:#[link(name = "android")]
./.cargo/git/checkouts/glutin-72cd277ab8976bc1/servo/src/android/ffi.rs:#[link(name = "android")]
./.cargo/git/checkouts/rust-azure-03a3575c638f374d/master/src/linkhack.rs:#[cfg(target_os = "android")]
./.cargo/git/checkouts/android-rs-glue-d59b5a2e8b246dc8/master/glue/src/ffi.rs:#[link(name = "android")]
./.cargo/git/checkouts/rust-mozjs-ab9bfe6e592bc16c/master/src/linkhack.rs:#[cfg(target_os = "android")]
./.cargo/git/checkouts/rust-harfbuzz-7808d225c82545c4/master/src/linkhack.rs:#[cfg(any(target_os = "macos", target_os = "android"))]
./.cargo/git/checkouts/glutin-666c1766c69e6584/master/src/android/ffi.rs:#[link(name = "android")]
./.cargo/git/checkouts/glutin-666c1766c69e6584/master/src/android/ffi.rs:#[link(name = "android")]
./support/android-rs-glue/glue/src/ffi.rs:#[link(name = "android")]
larsberg@ubuntu:~/servo$ 
```

I'm going to have to figure this out and possibly rebuild our B2G tarball, as we don't currently include libandroid.so and I'm not sure if we need the one ripped from the device itself or if one of the ones from the Android NDK will suffice. Ugh.

dmarcos commented:
@larsbergstrom Let me know if there's anything I can do to help

larsbergstrom commented:
@dmarcos Thanks for the offer, but I need to figure out how to address this on the B2G side - I'll make a PR against your PR once I have a fix :-)

larsbergstrom commented:
@dmarcos I have a PR for your PR that should fix the build here - https://github.com/dmarcos/servo/pull/2

larsbergstrom commented:
@bors-servo: r+

bors-servo commented:
:pushpin: Commit 0313a67 has been approved by `larsbergstrom`

<!-- @bors-servo r=larsbergstrom 0313a670c67977bbf43a7d3b82913f56d49714ae -->

bors-servo commented:
:hourglass: Testing commit 0313a67 with merge 014c96a...

dmarcos commented:
I Just pushed a new commit that includes @larsbergstrom https://github.com/dmarcos/servo/pull/2

larsbergstrom commented:
@bors-servo: r+

bors-servo commented:
:pushpin: Commit 0bc3cc2 has been approved by `larsbergstrom`

<!-- @bors-servo r=larsbergstrom 0bc3cc29bac89d1ecbd634e49c9cca90c680bb5a -->

bors-servo commented:
:hourglass: Testing commit 0bc3cc2 with merge fa0ec81...

bors-servo commented:
:broken_heart: Test failed - [linux1](http://build.servo.org/builders/linux1/builds/1471)

larsbergstrom commented:
Test failure:
```
failures:

---- webgl-context/clearcolor.html == webgl-context/clearcolor_ref.html stdout ----
	thread 'webgl-context/clearcolor.html == webgl-context/clearcolor_ref.html' panicked at 'rendering difference: /tmp/servo-reftest-000305-diff.png', ../../tests/reftest.rs:337

stack backtrace:
   1:     0x7f9569cc8812 - sys::backtrace::write::h754dbf5d2d3172ebPBA
   2:     0x7f9569cce3a2 - panicking::on_panic::hdb9e837ac064f8afiHJ
   3:     0x7f9569ca3ce9 - rt::unwind::begin_unwind_inner::h488386ed3236ccf5lnJ
   4:     0x7f9569ca4091 - rt::unwind::begin_unwind_fmt::hac94a0d8ab692f55WlJ
   5:     0x7f9569c274f9 - check_reftest::h38723b12a6d0bbdbqKa
                        at /home/servo/buildbot/slave/linux1/build/components/servo/<std macros>:9
   6:     0x7f9569c26799 - make_test::closure.6137
                        at ../../tests/reftest.rs:242
   7:     0x7f9569c26691 - thunk::Thunk<'a, (), R>::new::closure.6134
                        at /home/larsberg/rust/src/libstd/thunk.rs:27
   8:     0x7f9569c2655f - thunk::F.Invoke<A, R>::invoke::h18076791199441036107
                        at /home/larsberg/rust/src/libstd/thunk.rs:54
   9:     0x7f9569c5760b - thunk::F.Invoke<A, R>::invoke::h15694365567613239476
  10:     0x7f9569c4eec4 - thunk::F.Invoke<A, R>::invoke::h814591684781874558
  11:     0x7f9569c4f340 - rt::unwind::try::try_fn::h19251769324565050
  12:     0x7f9569cd84a8 - rust_try_inner
  13:     0x7f9569cd8495 - rust_try
  14:     0x7f9569c4facb - thunk::F.Invoke<A, R>::invoke::h5897503781166755732
  15:     0x7f9569cccde5 - sys::thread::thread_start::h04d056c6180f4496s8E
  16:     0x7f956939c181 - start_thread
  17:     0x7f9568cab47c - __clone
  18:                0x0 - <unknown>


failures:
    webgl-context/clearcolor.html == webgl-context/clearcolor_ref.html

test result: FAILED. 312 passed; 1 failed; 0 ignored; 0 measured
```

larsbergstrom commented:
@dmarcos On the upside, it appears the gonk build passed :-)

jdm commented:
Additionally,
```
/_mozilla/mozilla/interfaces.html
---------------------------------
FAIL Interfaces exposed on the window
```
That will need the new rendering context interface added to it. Sucks about the reftest again - @larsbergstrom what magic did you do in order to use the reftest analyzer last time?

dmarcos commented:
I don't know how to do with the ref test I added for this patch. It [passes locally](https://camo.githubusercontent.com/43cdecc3902cfdadbb8253568a21f107a766b24c/687474703a2f2f662e636c2e6c792f6974656d732f314f30583265317532673032314b326f317531512f496d616765253230323031352d30342d3135253230617425323031312e34382e3131253230414d2e706e67) on my OSX computer and [also on Linux](https://cloud.githubusercontent.com/assets/1323194/7170437/6cf49192-e3ce-11e4-82ca-45130dcf7a0f.png) on @ecoal95 machine

larsbergstrom commented:
The way to test the CSS WG reftests is to do:
```
./mach test-wpt _mozilla/css/abs_float_pref_width.html --log-raw=something.log
```

And then take something.log and put it on http://hoppipolla.co.uk/410/reftest-analyser-structured.xhtml, which will show you where the pixels differed.

But if this is just a normal old Servo reftest, there should be a /tmp/servo-reftest-000305-diff.png file locally that shows where the pixels differed.

jdm commented:
Ok, looks like someone will need to log in to the build machine and see what's up by checking out the png. I'll do that in a couple hours if nobody else beats me.

larsbergstrom commented:
I have the diff here:
![servo-reftest-000305-diff](https://cloud.githubusercontent.com/assets/475847/7211240/a66b43a2-e51e-11e4-8f4b-fba01df9b840.png)

But unfortunately not the baseline images, as they're overwritten by each new build :-/

dmarcos commented:
@larsbergstrom It looks that the pixels don't get painted on buildbot. This is what my machine produces:

http://f.cl.ly/items/3Z2a3p3G0Z1s071a0536/Image%202015-04-17%20at%202.31.47%20PM.png

dmarcos commented:
@jdm @larsbergstrom Do any of you have a access to Linux machine to run the test? 
```./mach run tests/ref/webgl-context/clearcolor.html```

We know that the test passes on Ubuntu and OSX.

ecoal95 commented:
@dmarcos I could try this monday from my University. The sooner the better, though...

dmarcos commented:
@jdm @larsbergstrom I found this opened issue https://github.com/servo/servo/issues/5643 Do you think it could be related?

jdm commented:
I wonder if we're performing layout and fetching the pixels before the GL operations occur? What if we make the 3d context mark the canvas element as dirty after every operation by calling document.content_changed?

dmarcos commented:
@jdm Just pushed a commit. It that what you meant?

larsbergstrom commented:
@dmarcos It's probably not that issue; that's specific to virtualbox (our servers are linode installs of ubuntu) and it would cause Servo to crash.

pcwalton commented:
I'm personally fine with landing this even with racy reftests. We know that our layers infrastructure is racy because we don't properly synchronize buffer updates and we don't have transactions implemented. I see random graphics races a lot and it was only a matter of time before they started happening on the bots too. (It's sad, but fixing all the races is going to be a lot of work and I don't think it should block WebGL landing.)

dmarcos commented:
@pcwalton I removed the commit that marks the node as dirty that @jdm suggested and pushed again. Should I somehow disable the ref test?

jdm commented:
I suppose we can mark it flaky_linux in order to ignore failures on linux, but so far it's failed every single time on the linux builder. That's not particularly encouraging.

dmarcos commented:
@jdm What do you want me to do? Do you want me to restore the commit that marks the node as dirty (https://github.com/dmarcos/servo/commit/fc06990)? I'm a bit lost a this point because I cannot make the test fail on neither OSX nor Linux.

pcwalton commented:
Perhaps we should delay this to discuss at the weekly meeting; then I can explain the issues and we can make a decision. Sorry about this :(

jdm commented:
@dmarcos In that case, let's go ahead and add the flaky_linux annotation to the reftest manifest and merge again. See http://mxr.mozilla.org/servo/source/tests/ref/basic.list#51 for a similar example.

dmarcos commented:
@jdm I installed Ubuntu on VMware on my mac and I don't see any pixels and I get an error in console that it's not very informative: ```[glutin] x error code=-78 major=-101 minor=34!```

dmarcos commented:
@jdm Actually I see the error on any site I open ```./mach run http://www.wikipedia.org``` gives me the same message

bors-servo commented:
:umbrella: The latest upstream changes (presumably #5559) made this pull request unmergeable. Please resolve the merge conflicts.

dmarcos commented:
@jdm @pcwalton Just rebased to keep this mergable. Just let me know if you have any ideas on why pixels don't show up on Linux.

jdm commented:
Let's just go ahead and add the flaky_linux annotation and merge this.

dmarcos commented:
@jdm Done!

jdm commented:
@bors-servo: r+

bors-servo commented:
:pushpin: Commit 4a29c45 has been approved by `jdm`

<!-- @bors-servo r=jdm 4a29c45c79f4d794a02844c604c0811b86ac66ea -->

bors-servo commented:
:hourglass: Testing commit 4a29c45 with merge b8789c2...

bors-servo commented:
:broken_heart: Test failed - [mac1](http://build.servo.org/builders/mac1/builds/1701)

dmarcos commented:
@jdm ```tests/wpt/metadata/2dcontext/the-canvas-state/canvas_state_restore_001.htm.ini``` was gone by mistake during rebase. I restored it and pushed again

pcwalton commented:
@bors-servo: r+

bors-servo commented:
:pushpin: Commit 2d9b197 has been approved by `pcwalton`

<!-- @bors-servo r=pcwalton 2d9b197c360de4b2b904677d5ab20d07a06c0312 -->

bors-servo commented:
:hourglass: Testing commit 2d9b197 with merge 6807555...

bors-servo commented:
:broken_heart: Test failed - [mac2](http://build.servo.org/builders/mac2/builds/1704)

jdm commented:
Heh, one last test failure to address:
```
/_mozilla/mozilla/interfaces.html
---------------------------------
FAIL Interfaces exposed on the window
```
You'll need to add the new 3d context interface to http://mxr.mozilla.org/servo/source/tests/wpt/mozilla/tests/mozilla/interfaces.html?force=1 .

dmarcos commented:
@jdm Just pushed the change. Here it is:

https://github.com/dmarcos/servo/commit/c82485874d1fe1db89a2dac61c6d612a75b8e1a1#diff-60cb8e34e787f23f73e256c83d621c8eR178

jdm commented:
@bors-servo: r+

bors-servo commented:
:pushpin: Commit c824858 has been approved by `jdm`

<!-- @bors-servo r=jdm c82485874d1fe1db89a2dac61c6d612a75b8e1a1 -->

bors-servo commented:
:hourglass: Testing commit c824858 with merge f31b813...

bors-servo commented:
:broken_heart: Test failed - [linux1](http://build.servo.org/builders/linux1/builds/1488)

jdm commented:
@bors-servo: retry

bors-servo commented:
:zap: Previous build results for [gonk](http://build.servo.org/builders/gonk/builds/663) are reusable. Rebuilding only [android](http://build.servo.org/builders/android/builds/1517), [linux1](http://build.servo.org/builders/linux1/builds/1488), [linux2](http://build.servo.org/builders/linux2/builds/1486), [mac1](http://build.servo.org/builders/mac1/builds/1705), [mac2](http://build.servo.org/builders/mac2/builds/1706)...

bors-servo commented:
:broken_heart: Test failed - [linux2](http://build.servo.org/builders/linux2/builds/1487)

jdm commented:
@bors-servo: retry
* #5764

dmarcos commented:
I think that with the latest patch that landed that test should pass (save and restore canvas state) That's the reason why the merge failed two runs ago. I had removed the ini fail because in my local machine the test was passing but failed on buildbot. Now the test passes as you would expect. @jdm What do we do?

bors-servo commented:
:hourglass: Testing commit c824858 with merge e4b620e...

jdm commented:
While true, we have merged other PRs since #5731 merged and they have not encountered that unexpected PASS result.

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/1519), [gonk](http://build.servo.org/builders/gonk/builds/665), [linux1](http://build.servo.org/builders/linux1/builds/1491), [linux2](http://build.servo.org/builders/linux2/builds/1489), [mac1](http://build.servo.org/builders/mac1/builds/1707), [mac2](http://build.servo.org/builders/mac2/builds/1709)

frewsxcv commented:
Congrats on this landing! :cake:

dmarcos commented:
I'm really happy that this finally landed :)
